File: test3.py
#!/usr/bin/python

from __future__ import division
from struct import *
import fcntl
import struct
import os
import errno

#
# Globals
#
DEVICE_PATH = '/dev/pubsub'
TYPE_NONE = 0
TYPE_PUB = 1
TYPE_SUB = 2

#
# Utilities for calculating the IOCTL command codes.
#
sizeof = {
    'byte': calcsize('c'),
    'signed byte': calcsize('b'),
    'unsigned byte': calcsize('B'),
    'short': calcsize('h'),
    'unsigned short': calcsize('H'),
    'int': calcsize('i'),
    'unsigned int': calcsize('I'),
    'long': calcsize('l'),
    'unsigned long': calcsize('L'),
    'long long': calcsize('q'),
    'unsigned long long': calcsize('Q')
}

_IOC_NRBITS = 8
_IOC_TYPEBITS = 8
_IOC_SIZEBITS = 14
_IOC_DIRBITS = 2

_IOC_NRMASK = ((1 << _IOC_NRBITS)-1)
_IOC_TYPEMASK = ((1 << _IOC_TYPEBITS)-1)
_IOC_SIZEMASK = ((1 << _IOC_SIZEBITS)-1)
_IOC_DIRMASK = ((1 << _IOC_DIRBITS)-1)

_IOC_NRSHIFT = 0
_IOC_TYPESHIFT = (_IOC_NRSHIFT+_IOC_NRBITS)
_IOC_SIZESHIFT = (_IOC_TYPESHIFT+_IOC_TYPEBITS)
_IOC_DIRSHIFT = (_IOC_SIZESHIFT+_IOC_SIZEBITS)

_IOC_NONE = 0
_IOC_WRITE = 1
_IOC_READ = 2

def _IOC(dir, _type, nr, size):
    if type(_type) == str:
        _type = ord(_type)
        
    cmd_number = (((dir)  << _IOC_DIRSHIFT) | \
        ((_type) << _IOC_TYPESHIFT) | \
        ((nr)   << _IOC_NRSHIFT) | \
        ((size) << _IOC_SIZESHIFT))

    return cmd_number

def _IO(_type, nr):
    return _IOC(_IOC_NONE, _type, nr, 0)

def _IOR(_type, nr, size):
    return _IOC(_IOC_READ, _type, nr, sizeof[size])

def _IOW(_type, nr, size):
    return _IOC(_IOC_WRITE, _type, nr, sizeof[size])

def main():
    print("start")
    """Test the device driver"""
    #
    # Calculate the ioctl cmd number
    #
    MY_MAGIC = 'r'
    SET_TYPE = _IO(MY_MAGIC, 0)
    GET_TYPE = _IO(MY_MAGIC, 1)

    # NATHAN

    f1 = os.open(DEVICE_PATH, os.O_RDWR)
    # Check we don't have a type
    assert (fcntl.ioctl(f1, GET_TYPE) == TYPE_NONE)

    # Make us a publisher
    fcntl.ioctl(f1, SET_TYPE, TYPE_PUB)
    # Check we have a type
    assert (fcntl.ioctl(f1, GET_TYPE) == TYPE_PUB)
    
    f2 = os.open(DEVICE_PATH, os.O_RDWR)
    # Check we don't have a type
    assert (fcntl.ioctl(f2, GET_TYPE) == TYPE_NONE)

    f3 = os.open(DEVICE_PATH, os.O_RDWR)
    # Check we don't have a type
    assert (fcntl.ioctl(f3, GET_TYPE) == TYPE_NONE)

    # Make us a publisher
    fcntl.ioctl(f2, SET_TYPE, TYPE_SUB)
    # Check we have a type
    assert (fcntl.ioctl(f2, GET_TYPE) == TYPE_SUB)
    # Make us a publisher
    fcntl.ioctl(f3, SET_TYPE, TYPE_SUB)
    # Check we have a type
    assert (fcntl.ioctl(f3, GET_TYPE) == TYPE_SUB)

    ret = os.write(f1, "x"*700)
    assert (ret == 700) # returns bytes written

    ret = os.read(f2, 700)
    assert (len(ret) == 700)

    ret = os.read(f3, 700)
    assert (len(ret) == 700)

    ret = os.write(f1, "x"*700)
    assert (ret == 700) # returns bytes written

    ret = os.read(f2, 500)
    assert (len(ret) == 500)

    ret = os.read(f2, 300)
    assert (len(ret) == 200)



    # # We should not be able to write
    # try:
    #     ret = os.write(f1, "X" * 1050)
    #     assert False # We should not reach this line
    # except OSError, e:
    #      # Check the correct error was raised
    #     assert (e.errno == errno.EINVAL)

    
    # ret = os.write(f1, "x"*900)
    # assert (ret == 900) # returns bytes written

    # try:
    #     ret = os.write(f1, "x"*101)
    #     print(ret)
    #     assert False # We should not reach this line
    # except OSError, e:
    #      # Check the correct error was raised
    #     assert (e.errno == errno.EAGAIN)


    # f2 = os.open(DEVICE_PATH, os.O_RDWR)
    # # Check we don't have a type
    # assert (fcntl.ioctl(f2, GET_TYPE) == TYPE_NONE)

    # f3 = os.open(DEVICE_PATH, os.O_RDWR)
    # # Check we don't have a type
    # assert (fcntl.ioctl(f3, GET_TYPE) == TYPE_NONE)

    # # Make us a publisher
    # fcntl.ioctl(f2, SET_TYPE, TYPE_SUB)
    # # Check we have a type
    # assert (fcntl.ioctl(f2, GET_TYPE) == TYPE_SUB)
    # # Make us a publisher
    # fcntl.ioctl(f3, SET_TYPE, TYPE_SUB)
    # # Check we have a type
    # assert (fcntl.ioctl(f3, GET_TYPE) == TYPE_SUB)

    # ret = os.read(f2, 1000)
    # print("ret: ",len(ret))
    # assert (len(ret) == 900)

    # try:
    #     ret = os.read(f2, 10)
    #     assert False # We should not reach this line
    # except OSError, e:
    #      # Check the correct error was raised
    #     assert (e.errno == errno.EAGAIN)

    # ret = os.read(f3, 500)
    # assert (len(ret) == 500)

    # ret = os.read(f3, 1000)
    # assert (len(ret) == 400)

    # try:
    #     ret = os.read(f3, 10)
    #     assert False # We should not reach this line
    # except OSError, e:
    #      # Check the correct error was raised
    #     assert (e.errno == errno.EAGAIN)

    # ret = os.write(f1, "x"*101)
    # assert (ret == 101) # returns bytes written

    # ret = os.read(f2, 200)
    # assert (len(ret) == 101)

    # ret = os.read(f3, 200)
    # assert (len(ret) == 101)

    # try:
    #     ret = os.read(f2, 10)
    #     assert False # We should not reach this line
    # except OSError, e:
    #      # Check the correct error was raised
    #     assert (e.errno == errno.EAGAIN)

    # try:
    #     ret = os.read(f3, 10)
    #     assert False # We should not reach this line
    # except OSError, e:
    #      # Check the correct error was raised
    #     assert (e.errno == errno.EAGAIN)

    os.close(f1)
    os.close(f2)
    os.close(f3)

    
if __name__ == '__main__':
    main()
    

File: test3.c
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <errno.h>
#include <string.h>
#include <time.h>

#define DEVICE_PATH "/dev/pubsub"
#define BUFFER_SIZE 1000
#define PUB_TYPE 1
#define SUB_TYPE 2
#define SET_TYPE  _IO('r', 0)
#define GET_TYPE  _IO('r', 1)

#define GREEN "\033[32m"
#define RED "\033[31m"
#define YELLOW "\033[33m"
#define RESET "\033[0m"

// Test counter
static int test_count = 0;
static int pass_count = 0;

void assert_test(int condition, const char *test_description) {
    test_count++;
    if (condition) {
        pass_count++;
        printf(GREEN "[PASS %d] " RESET "%s\n", test_count, test_description);
    } else {
        printf(RED "[FAIL %d] %s\n", test_count, test_description);
        printf("Error: %s (errno=%d)\n" RESET, strerror(errno), errno);
        exit(EXIT_FAILURE);
    }
}

void test_suite_banner(const char *name) {
    printf(YELLOW "\n=== %s ===\n" RESET, name);
}

void test_type_setting() {
    test_suite_banner("Testing Type Setting and Validation");
    
    int fd = open(DEVICE_PATH, O_RDWR);
    assert_test(fd >= 0, "Open device");

    // Test initial state
    int type = ioctl(fd, GET_TYPE, 0);
    assert_test(type == 0, "Initial type should be TYPE_NONE");

    // Test invalid type
    int ret = ioctl(fd, SET_TYPE, 999);
    assert_test(ret == -1 && errno == EINVAL, "Setting invalid type should fail");

    // Test valid pub type
    ret = ioctl(fd, SET_TYPE, PUB_TYPE);
    assert_test(ret == 0, "Setting valid PUB_TYPE");

    // Verify type was set
    type = ioctl(fd, GET_TYPE, 0);
    assert_test(type == PUB_TYPE, "Verify type is PUB_TYPE");

    // Try to change type (should fail)
    ret = ioctl(fd, SET_TYPE, SUB_TYPE);
    assert_test(ret == -1 && (errno == EINVAL || errno == EPERM), 
                "Changing type should fail");

    close(fd);
}

void test_buffer_limits() {
    test_suite_banner("Testing Buffer Limits");
    
    int fd = open(DEVICE_PATH, O_RDWR);
    assert_test(fd >= 0, "Open device");

    // Set as publisher
    assert_test(ioctl(fd, SET_TYPE, PUB_TYPE) == 0, "Set as publisher");

    // Test writing exactly BUFFER_SIZE
    char *big_buffer = malloc(BUFFER_SIZE);
    memset(big_buffer, 'A', BUFFER_SIZE);
    
    int ret = write(fd, big_buffer, BUFFER_SIZE);
    assert_test(ret == BUFFER_SIZE, "Write exactly BUFFER_SIZE bytes");

    // Test writing more than BUFFER_SIZE
    ret = write(fd, big_buffer, BUFFER_SIZE + 1);
    assert_test(ret == -1 && errno == EINVAL, "Write > BUFFER_SIZE should fail");

    free(big_buffer);
    close(fd);
}

void test_pub_sub_interaction() {
    test_suite_banner("Testing Publisher-Subscriber Interaction");

    int pub_fd = open(DEVICE_PATH, O_RDWR);
    int sub_fd = open(DEVICE_PATH, O_RDWR);
    
    // Setup publisher and subscriber
    assert_test(ioctl(pub_fd, SET_TYPE, PUB_TYPE) == 0, "Set publisher");
    assert_test(ioctl(sub_fd, SET_TYPE, SUB_TYPE) == 0, "Set subscriber");

    // Test reading when no data available
    char read_buf[100];
    int ret = read(sub_fd, read_buf, sizeof(read_buf));
    assert_test(ret == -1 && errno == EAGAIN, "Read with no data should return EAGAIN");

    // Write test data
    const char *test_msg = "Test Message";
    size_t msg_len = strlen(test_msg);
    ret = write(pub_fd, test_msg, msg_len);
    assert_test(ret == msg_len, "Write test message");

    // Read data
    memset(read_buf, 0, sizeof(read_buf));
    ret = read(sub_fd, read_buf, sizeof(read_buf));
    assert_test(ret == msg_len, "Read should return written length");
    assert_test(memcmp(read_buf, test_msg, msg_len) == 0, "Data integrity check");

    close(pub_fd);
    close(sub_fd);
}

void test_error_conditions() {
    test_suite_banner("Testing Error Conditions");

    int fd = open(DEVICE_PATH, O_RDWR);
    assert_test(fd >= 0, "Open device");

    // Test read/write without setting type
    char buffer[10] = "test";
    int ret = write(fd, buffer, sizeof(buffer));
    assert_test(ret == -1 && errno == EPERM, "Write without type should fail");

    ret = read(fd, buffer, sizeof(buffer));
    assert_test(ret == -1 && errno == EPERM, "Read without type should fail");

    // Set as subscriber and try to write
    assert_test(ioctl(fd, SET_TYPE, SUB_TYPE) == 0, "Set as subscriber");
    ret = write(fd, buffer, sizeof(buffer));
    assert_test(ret == -1 && errno == EPERM, "Write as subscriber should fail");

    // Test invalid ioctl command
    ret = ioctl(fd, 9999, 0);
    assert_test(ret == -1 && errno == ENOTTY, "Invalid ioctl command should fail");

    close(fd);

    // Test operations on closed fd
    ret = write(fd, buffer, sizeof(buffer));
    assert_test(ret == -1 && errno == EBADF, "Write to closed fd should fail");
}

void stress_test() {
    test_suite_banner("Testing stress_test ");

    int pub_fd = open(DEVICE_PATH, O_RDWR);
    int sub_fd = open(DEVICE_PATH, O_RDWR);

    assert_test(ioctl(pub_fd, SET_TYPE, PUB_TYPE) == 0, "Set publisher");
    assert_test(ioctl(sub_fd, SET_TYPE, SUB_TYPE) == 0, "Set subscriber");

    char buffer[BUFFER_SIZE];
    const int iterations = 1000;

    // Rapid write/read cycles
    int i;
    for (i = 0; i < iterations; i++) {
        snprintf(buffer, sizeof(buffer), "Message %d", i);
        int write_ret = write(pub_fd, buffer, strlen(buffer));
        assert_test(write_ret >= 0, "Stress write");

        char read_buf[BUFFER_SIZE];
        int read_ret = read(sub_fd, read_buf, sizeof(read_buf));
        assert_test(read_ret >= 0, "Stress read");
        assert_test(memcmp(read_buf, buffer, strlen(buffer)) == 0, "Stress data integrity");
    }

    close(pub_fd);
    close(sub_fd);
}

int main() {
    srand(time(NULL));
    printf("Starting Comprehensive PubSub Driver Test\n");
    printf("=======================================\n");

    test_type_setting();
    test_buffer_limits();
    test_pub_sub_interaction();
    test_error_conditions();
    stress_test();

    printf(GREEN "\nTest Summary: %d/%d tests passed\n" RESET, pass_count, test_count);
    return 0;
}
